community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/pipes/DistinctVarLengthExpandPipe.scala

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v3_2.pipes
+
+import org.neo4j.collection.primitive.{Primitive, PrimitiveLongObjectMap}
+import org.neo4j.cypher.internal.compiler.v3_2.ExecutionContext
+import org.neo4j.cypher.internal.compiler.v3_2.executionplan.{Effects, ReadsAllNodes, ReadsAllRelationships}
+import org.neo4j.cypher.internal.compiler.v3_2.planDescription.InternalPlanDescription.Arguments.ExpandExpression
+import org.neo4j.cypher.internal.frontend.v3_2.symbols._
+import org.neo4j.cypher.internal.frontend.v3_2.{InternalException, SemanticDirection}
+import org.neo4j.graphdb.{Node, Relationship}
+
+case class DistinctVarLengthExpandPipe(
+                                 source: Pipe,
+                                 fromName: String,
+                                 relName: String,
+                                 toName: String,
+                                 types: LazyTypes,
+                                 dir: SemanticDirection,
+                                 min: Int,
+                                 max: Int,
+                                 filteringStep: VarLengthPredicate = VarLengthPredicate.NONE)
+                                (val estimatedCardinality: Option[Double] = None)
+                                (implicit pipeMonitor: PipeMonitor) extends PipeWithSource(source, pipeMonitor) with RonjaPipe {
+  self =>
+
+  trait State {
+    // Note that the Node part here can be null.
+    // This code is used in a hot spot, and avoiding object creation is important.
+    def next(): (State, ExecutionContext)
```

Comment:
It looks like new states get created all the time so they could also just contain the ExecutionContext which might save us one object allocation for the tuple.